### PR TITLE
Merge with original repo

### PR DESCRIPTION
**Pull request by adiramani**
- Adding github actions to add some forms of automated testing when new pull requests are made.
- This action will run the model with predetermined intervention history and forward simulation. Using MF prevalence data from previous runs, it will visualize the new models runs with a previously correct version of the python model, and the r model, so that people can compare their results.
- In theory, before creating a pull request, this test should be run locally to make sure nothing breaks.

See the most recent comment from "https://github.com/adiramani/P_EPIONCHO-IBM/pull/6" to see what the output will be.

### CODE REVIEW

**Comment by tlestang** [comment]
It looks like this brings two duplicate commits that were already merged in #70

**Comment by tlestang** [comment]
A lot of work! Thank you @adiramani 

I just had a rapid glance, but it looks like the 3 test scripts are very similar. I'd advise that you (by order of priority):
- Include module level docstring that describe what the test tests (and how it differs from others)
- Factor the duplicated code into separate modules to reduce each testing module script to code that is really specific to it. For instance I think you could extract most of the plotting logic into a single plotting module. 

But tests - even complicated - are better than no tests. So if you don't have a lot of time right now, I'm happy to merge this in and you or someone else can revisit this later.

**Comment by adiramani** [comment]
I can definitely improve the tests (and I am working on gathering all the use cases). For now, since I am the only person working on the model, it is fine to leave this unmerged, since I can test my changes locally first, and we can wait to have a much more refined testing module merged. If/when someone else starts to make changes, we can push this change (assuming I haven't improved the testing by then)

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+              '**Please look at the Comparison Plot and verify that your changes are valid.**\n' +
+              'Access the workflow information:\n' +
+              '| Type | Link |\n' +
+              '| --- | --- |\n' +
+              `| Workflow | ${workflow_url} |\n`+
+              `| Comparison Plot | ${artifact_url} |\n` +
+              `| Annual MOX Sequelae Plots | ${artifact_url2} |\n` +
+              `| Biannual MOX Sequelae Plots | ${artifact_url3} |\n` +
+              `| Quadnnual MOX Sequelae Plots | ${artifact_url4} |\n` +
+              `To run this test locally, run the code in the [examples/testing_modules/test_new_changes.py](${test_file}) file.`
```
Comment (tlestang) [inline .github/workflows/main.yml:131]: `test_file` is a path to `examples/test_new_changes.py` and not `examples/testing_modules/test_new_changes.py`

File: examples/testing_modules/test_new_changes_hdf5.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            with_age_groups=True,
+            prevalence=True,
+            mean_worm_burden=True,
+            prevalence_OAE=True,
+            intensity=True,
+            with_sequela=True,
+            with_pnc=True,
+        )
+
+    return (run_data, run_data_age)
```
Comment (tlestang) [inline examples/testing_modules/test_new_changes_hdf5.py:91]: What is the difference between `run_data_age` and `run_data`?

**Review by tlestang — COMMENTED** [review]
Thanks @adiramani for all that work. I've got two comments:

- It looks to me like modules in the `examples/testing_modules` directory are doing the same thing as `examples/testing_new_changes.py`. Is there a need to keep both? My guess is that the former comes from refactoring the latter, but hard to tell.
- The test run several simulations across a pool of workers. Is there specific code that can only be tested by running multiple processes in parallel. I think the testing code could be a lot simpler without having to wrangle `run_sim`, `run_simulation`, `runSim`, workers and `process_map`s.

File: examples/testing_modules/test_new_changes_hdf5.py (hunk 2 above)
Comment (adiramani) [reply examples/testing_modules/test_new_changes_hdf5.py:91]: run_data_age is grouped by age in 1 year age bins, vs run_data is grouped for all ages. The outputs of the test use both, with run_data used to verify that the age grouped data adds up to the correct values.

File: .github/workflows/main.yml (hunk 1 above)
Comment (adiramani) [reply .github/workflows/main.yml:131]: Ah good catch, I shall change that

**Comment by adiramani** [comment]
> Thanks @adiramani for all that work. I've got two comments:
> 
> * It looks to me like modules in the `examples/testing_modules` directory are doing the same thing as `examples/testing_new_changes.py`. Is there a need to keep both? My guess is that the former comes from refactoring the latter, but hard to tell.

No there is not reason to keep both, it was just an old artifact I forgot to remove, so good catch.

> * The test run several simulations across a pool of workers. Is there specific code that can only be tested by running multiple processes in parallel. I think the testing code could be a lot simpler without having to wrangle `run_sim`, `run_simulation`, `runSim`, workers and `process_map`s.

I should be able to simplify the total number of function calls, but we do need to test the processes in parallel, as we need at least 50 runs to get a viable "mean prevalence" to visualize

**Comment by tlestang** [comment]
> but we do need to test the processes in parallel, as we need at least 50 runs to get a viable "mean prevalence" to visualize

Ah okay I see. That probably justifies the complexity then.

> no reason to keep both, it was just an old artifact I forgot to remove, so good catch.

I'm good to merge this as soon as you can remove this obsolete files.

**Comment by adiramani** [comment]
> > but we do need to test the processes in parallel, as we need at least 50 runs to get a viable "mean prevalence" to visualize
> 
> Ah okay I see. That probably justifies the complexity then.
> 
> > no reason to keep both, it was just an old artifact I forgot to remove, so good catch.
> 
> I'm good to merge this as soon as you can remove this obsolete files.

Just removed the obsolete files and edited the text in `main.yml`

**Pull request merged**